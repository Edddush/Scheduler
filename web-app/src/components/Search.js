/*eslint no-useless-escape: 0*/
import React, { useState, useEffect } from "react";
import { prettifyMeetingInformation } from "./Util";

//Search functionality
function Search({ setCurrentPage,
  setSearchResults,
  searchResults,
  searchInput,
  setSearchInput,
  setLoading
}) {
  //Initializing states
  const [errorMessage, setErrorMessage] = useState("");

  // detect if given string is empty, undefined of blank
  function blank(input) {
    if (typeof input === 'undefined') {
      return true;
    }

    return /^\s*$/.test(input);
  }

  useEffect(() => {
    setLoading(true);

    //Making a post request to flask
    (async () => {
      let valid_title = searchInput["courseTitle"];
      let course = searchInput["courseCode"];
      let dept = searchInput["subject"];

      if (blank(course) && blank(valid_title) && blank(dept)) {
        setLoading(false);
        setSearchResults([]);
        setErrorMessage("");
        return;
      }

      //fetching the search results from flask
      let result = await fetch(
        "api/courseSearch?" + new URLSearchParams(searchInput),
        {
          headers: {
            accepts: "application/json",
          },
        }
      );

      //catch the error if it occurred in flask
      if (result.status === 404) {
        setSearchResults([]);
        setErrorMessage("");
        setLoading(false);
        return;
      }

      //Format the return from flask
      result = await result.json();

      //Creating a row to display for each course/section in results
      let rows = [];
      if (result) {
        for (const course of result) {
          rows.push({
            ...course,
            meeting_info_string: prettifyMeetingInformation(
              course.meeting_information
            ),
          });
        }

        setSearchResults(rows);
        setErrorMessage("");
      } else {
        setErrorMessage("Error Occured During Search!");
      }
      setLoading(false);
    })();

    return;
  }, [searchInput]);
  /*   
    define behavior when input is changed
    update searchResults and reset pagination 
  */
  const handleChange = (e) => {
    const result = searchResults.filter(section => {
      (section.toString().toLowerCase().indexOf(e.target.value.toLowerCase())) > -1
    });

    setSearchInput((prev) => {
      return {
        ...prev,
        [e.target.id]: e.target.value,
      };
    });

    setCurrentPage(1);
    setSearchResults(result);
  };

  // delay callback by a given amount of time to allow full input 
  function debounce(callBack, delay = 500) {
    let timer = 0;

    return (...args) => {
      clearTimeout(timer)
      timer = setTimeout(() => {
        callBack(...args)
      },
        delay
      )
    }
  }

  //Called when input changes
  const debounceChange = debounce(handleChange);

  const inputIds = [
    { id: "courseTitle", placeholder: "Enter a Course" },
    { id: "subject", placeholder: "Enter a Subject Code" },
    { id: "courseCode", placeholder: "Enter a Course Code" }];

  //Returning the component
  return (
    <div className="card-body" id="search">
      <div id="search-input-group">
        {inputIds.map(({ id, placeholder }) => (
          <input
            key={id}
            id={id}
            type="text"
            className="search-box"
            placeholder={placeholder}
            onChange={debounceChange}
          />
        ))}
        { }
      </div>
      <h4>{errorMessage}</h4>
    </div>
  );
}

export default Search;
