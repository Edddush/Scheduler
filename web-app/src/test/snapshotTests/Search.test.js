import React from 'react';
import renderer from 'react-test-renderer';
import Search from '../../components/Search';
import ScheduleProvider from '../../context/ScheduleContext'

it('renders with no props', () => {
  const tree = renderer
    .create(
      <ScheduleProvider>
        <Search />
      </ScheduleProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

