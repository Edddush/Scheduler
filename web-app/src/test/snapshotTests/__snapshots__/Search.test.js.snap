// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with no props 1`] = `
<div
  className="card-body"
  id="search"
>
  <div
    id="search-input-group"
  >
    <input
      className="search-box"
      id="courseTitle"
      onChange={[Function]}
      placeholder="Enter a Course"
      type="text"
    />
    <input
      className="search-box"
      id="subject"
      onChange={[Function]}
      placeholder="Enter a Subject Code"
      type="text"
    />
    <input
      className="search-box"
      id="courseCode"
      onChange={[Function]}
      placeholder="Enter a Course Code"
      type="text"
    />
  </div>
  <h4 />
</div>
`;
